vpr/src/analytical_place/partial_legalizer.cpp

```diff
@@ -0,0 +1,1104 @@
+        }
+        VTR_LOG("\n");
+    }
+    VTR_LOG("\n");
+}
+
+void FlowBasedLegalizer::legalize(PartialPlacement &p_placement) {
+    VTR_LOGV(log_verbosity_ >= 10, "Running Flow-Based Legalizer\n");
+
+    // Reset the bins from the previous iteration and prepare for this iteration.
+    reset_bins();
+    // Import the partial placement into bins.
+    import_placement_into_bins(p_placement);
+    // Verify that the placement was imported correctly.
+    VTR_ASSERT_SAFE(verify_bins());
+
+    // Print the number of blocks in each bin visually before spreading.
+    if (log_verbosity_ >= 15) {
+        VTR_LOG("Bin utilization prior to spreading:\n");
+        print_flow_based_bin_grid(tile_bin_, bins_);
+    }
+
+    // Print the status header to make printing the status clearer.
+    if (log_verbosity_ >= 10) {
+        print_flow_based_legalizer_status_header();
+    }
+
+    // Run the flow-based spreader.
+    size_t flowBasedIter = 0;
+    while (true) {
+        // If we hit the maximum number of iterations, break.
+        if (flowBasedIter >= max_num_iterations_) {
+            VTR_LOGV(log_verbosity_ >= 10,
+                     "Flow-Based legalizer hit max iteration limit.\n");
+            break;
+        }
+        // If there are no overfilled bins, no more work to do.
+        if (overfilled_bins_.empty()) {
+            VTR_LOGV(log_verbosity_ >= 10,
+                     "Flow-Based legalizer has no overfilled tiles. No further spreading needed.\n");
+            break;
+        }
+        // Compute the max movement.
+        double psi = computeMaxMovement(flowBasedIter);
+        // Get the overfilled bins and sort them in increasing order of supply.
+        // We take the manhatten (L1) norm here since we only care about the total
+        // amount of overfill in each dimension. For example, a bin that has a
+        // supply of <1, 1> is just as overfilled as a bin of supply <0, 2>.
+        // The standard L2 norm would give more weigth to <0, 2>.
+        // NOTE: Although the supply should always be non-negative, we still
+        //       take the absolute value in the norm for completeness.
+        // TODO: This is a guess. Should investigate other norms.
+        std::vector<LegalizerBinId> overfilled_bins_vec(overfilled_bins_.begin(), overfilled_bins_.end());
+        std::sort(overfilled_bins_vec.begin(), overfilled_bins_vec.end(), [&](LegalizerBinId a, LegalizerBinId b) {
+            return bins_[a].supply.manhatten_norm() < bins_[b].supply.manhatten_norm();
+        });
+        // Get the paths to flow blocks from the overfilled bins to the under
+        // filled bins and flow the blocks.
+        for (LegalizerBinId bin_id : overfilled_bins_vec) {
+            // Get the list of candidate paths based on psi.
+            //  NOTE: The paths are sorted by increasing cost within the
+            //        getPaths method.
+            std::vector<std::vector<LegalizerBinId>> paths = get_paths(bin_id,
```

Comment:
Could put a comment saying "Get multiple candidate paths, each of which is a vector of the legalizerBins traversed."